### PR TITLE
[IMP] point_of_sale: Send ticket by email.

### PR DESCRIPTION
**Pull request by rhe-odoo**
Send the receipt to the customer instead of auto printing.

Description of the issue/feature this PR addresses:
Be able to send a PoS Ticket by email. 

Current behavior before PR:
Whenever you make a payment in the pos, you'll automatically print the ticket.

Desired behavior after PR is merged:
When you make a payment, you now have the possibility to send the ticket to the customer by mail.
If you do, the ticket won't be auto printed but you'll have the possibility to print it anyway.

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pimodoo** [comment]
@rhe-odoo explain more in commit message and put TASK-ID

File: addons/point_of_sale/models/pos_order.py [hunk 1]
```diff
@@ -558,6 +558,18 @@ def refund(self):
             'target': 'current',
         }
 
+    def action_receipt_to_customer(self, receipt, cashier_id, client):
+        template_obj = self.env['mail.mail']
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_order.py]: you didn't do for loop because it works on only one record. so you need to do self.ensure_one()

File: addons/point_of_sale/models/pos_order.py [hunk 2]
```diff
@@ -558,6 +558,18 @@ def refund(self):
...
+        message = "<p>Dear %s,<br/>Here is your electronic ticket from the %s. </p>" % (client['name'], self.id)
+        template_data = {
+            'subject': 'Receipt %s' % self.id,
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_order.py]: you probably should use self.pos_reference as it is the receipt reference

File: addons/point_of_sale/models/pos_order.py [hunk 3]
```diff
@@ -558,6 +558,18 @@ def refund(self):
...
+            'body_html': message + '<img src="data:image/jpeg;base64,%s"/>' % receipt,
+            'email_from': self.env['res.users'].search([('id', '=', cashier_id)]).email,
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_order.py]: check with laura, but we probably don't want to use the email of the user, but the mail of the admin or something like this

File: addons/point_of_sale/static/src/js/screens.js [hunk 4]
```diff
@@ -2165,13 +2209,11 @@ var PaymentScreenWidget = ScreenWidget.extend({
 
         return true;
     },
-
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/screens.js]: why did you remove this line?

File: addons/point_of_sale/models/pos_order.py [hunk 5]
```diff
@@ -558,6 +558,18 @@ def refund(self):
             'target': 'current',
         }
 
+    def action_receipt_to_customer(self, receipt, cashier_id, client):
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_order.py]: if the order is already server side, cashier_id -> self.user_id and client -> self.partner_id

File: addons/point_of_sale/static/src/js/screens.js [hunk 6]
```diff
@@ -1765,6 +1767,25 @@ var ReceiptScreenWidget = ScreenWidget.extend({
     render_receipt: function() {
         this.$('.pos-receipt-container').html(QWeb.render('OrderReceipt', this.get_receipt_render_env()));
     },
+    send_receipt_to_customer: function() {
+        if(this.pos.get_order().is_to_email()){
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/screens.js]: What if we don't have internet access, what if the order is not on the server?

File: addons/point_of_sale/models/pos_order.py (hunk 3 above)
Comment (legalsylvain) [reply addons/point_of_sale/models/pos_order.py]: In most case, the email of the current company is the more relevant. Something like contact@mycompany.com
Admin email will be wrong, specially in a multi company context.

File: addons/point_of_sale/static/src/js/models.js [hunk 7]
```diff
@@ -2764,6 +2764,13 @@ exports.Order = Backbone.Model.extend({
     is_to_invoice: function(){
         return this.to_invoice;
     },
+    /* ---- Email --- */
+    set_to_email: function(to_email) {
+        this.to_email = to_email;
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/models.js]: don't forget to put 'this.to_email' in the initialize of the model

File: addons/point_of_sale/models/pos_order.py [hunk 8]
```diff
@@ -604,6 +604,19 @@ def refund(self):
             'target': 'current',
         }
 
+    @api.model
+    def action_receipt_to_customer(self, fields):
```
Comment (pimodoo) [inline addons/point_of_sale/models/pos_order.py]: don't put argument 'fields', just put action_receipt_to_customer(self, name, email, receipt):

File: addons/point_of_sale/static/src/js/screens.js [hunk 9]
```diff
@@ -2146,6 +2181,17 @@ var PaymentScreenWidget = ScreenWidget.extend({
             }
         }
 
+        if (order.is_to_email() && !order.get_client().email) {
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/screens.js]: you also need to check that there is a client

File: addons/point_of_sale/static/src/js/screens.js [hunk 10]
```diff
@@ -2192,8 +2238,18 @@ var PaymentScreenWidget = ScreenWidget.extend({
...
+            var ordered = this.pos.push_order(order);
+            ordered.then(function() {
+                self.gui.show_screen('receipt');
+            });
+            ordered.catch(function(value) {
+                order.set_to_email(false);
+                self.gui.show_screen('receipt');
+                self.gui.show_popup('error',{
+                    'title': "Error: no internet connection",
+                    'body':  "There is no internet connection, impossible to send the email. The ticket will be print.",
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/screens.js]: 'will be printed'. Not sure about the message, are you sure the ticket will be printed, or the cashier will have to 'unthick' the checkbox send by mail to print it.

File: addons/point_of_sale/static/src/js/screens.js [hunk 11]
```diff
@@ -2192,8 +2238,18 @@ var PaymentScreenWidget = ScreenWidget.extend({
                 self.gui.show_screen('receipt');
             });
         } else {
-            this.pos.push_order(order);
-            this.gui.show_screen('receipt');
+            var ordered = this.pos.push_order(order);
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/screens.js]: you only managed the case push order, you also need to manage when there is an invoice, by modifying your server function with:  action_receipt_to_customer(self, name, email, receipt, invoice_id=False), and give it an invoice id in this specific case and add the invoice pdf to the email containing receipt.

File: addons/point_of_sale/static/src/js/models.js [hunk 12]
```diff
@@ -899,6 +899,7 @@ exports.PosModel = Backbone.Model.extend({
                                 // The order has been pushed separately in batch when
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/models.js]: if (order_server_id.length && !order.is_to_email())
else if order_server_id.length -> resolve
else -> reject

File: addons/point_of_sale/static/src/js/screens.js [hunk 13]
```diff
@@ -1632,6 +1633,30 @@ var ReceiptScreenWidget = ScreenWidget.extend({
             this.lock_screen(false);
         }
     },
+    send_receipt_to_customer: function(order, orderServerId) {
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/screens.js]: mover send_receipt_to_customer -> PaymentScreen
remove order -> this.pos.get_order
rename orderServerId -> invoice_server_id

File: addons/point_of_sale/static/src/js/screens.js [hunk 14]
```diff
@@ -2303,13 +2352,27 @@ var PaymentScreenWidget = ScreenWidget.extend({
 
             invoiced.catch(this._handleFailedPushForInvoice.bind(this, order, false));
 
-            invoiced.then(function () {
+            invoiced.then(function (value) {
                 self.invoicing = false;
+                screen.send_receipt_to_customer(order, value.orderId[0]);
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/screens.js]: value.orderId[0] || false

File: addons/point_of_sale/static/src/js/screens.js [hunk 15]
```diff
@@ -1632,6 +1633,30 @@ var ReceiptScreenWidget = ScreenWidget.extend({
...
+            }
+
+            var receipt = QWeb.render('OrderReceipt', data);
+            var printer = new Printer();
+
+            printer.htmlToImg(receipt).then(function(ticket) {
+                rpc.query({
+                    model: 'pos.order',
+                    method: 'action_receipt_to_customer',
+                    args: [order.get_name(), order.get_client(), ticket, orderServerId? orderServerId : false],
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/screens.js]: only invoice_server_id (set to false when called)

File: addons/point_of_sale/static/src/js/models.js [hunk 16]
```diff
@@ -899,6 +899,7 @@ exports.PosModel = Backbone.Model.extend({
...
                                 // the connection came back.
                                 // The user has to go to the backend to print the invoice
+                                resolveInvoiced({orderId: order_server_id});
```
Comment (pimodoo) [inline addons/point_of_sale/static/src/js/models.js]: resolveInvoiced(order_server_id)

**Comment by pimodoo** [comment]
robodoo r+ rebase-ff

**Comment by pimodoo** [comment]
robodoo r+
